chore(vrl): configure Netlify ignore builds options

jonathanpv opened the pull request:
Closes https://github.com/vectordotdev/vector/issues/14950

This pr is meant to address the web playground being triggered on every change, when we would only like it to run
and deploy on changes made to lib/vrl/stdlib or lib/vrl/web-playground

jonathanpv commented:
I'd like to discuss whether we should add listening to changes on `lib/vrl/core`, `lib/vrl/compiler`, `lib/vrl/vrl`

I believe we wouldn't need to as the features we specify on the webplayground cargo.toml seem to read from the `lib/vrl/stdlib` crate.

Unsure if changes to the other crates mentioned would ever cause the vrl playground to fail. 

@JeanMertz @fuchsnj thoughts on this?

spencergilbert commented:
Aside from their thoughts on watching the other libs, we'll want to have this merge in your fix for `js` to make sure builds are happy.

spencergilbert reviewed (approved):
LGTM, up to y'all if you want to build on other VRL directories

arshiyasolei commented on lib/vrl/web-playground/netlify.toml:
```diff
@@ -2,18 +2,16 @@
 
   # Default build command.
-  command = "cargo install --version 0.10.3 wasm-pack && wasm-pack build --target web --out-dir public"
+  command = "cargo install --force --version 0.10.3 wasm-pack && wasm-pack build --target web --out-dir public"
```
I wonder if there's a way to suppress that warning instead of ```--force```

jszwedko replied:
This seems ok. The downside is that it'll install it even if it already is installed. In some other places, we have guards that check if the given tool is already installed and skip if now. Like this:

https://github.com/vectordotdev/vector/blob/d21c420e33ccf603dd69bb9b20d9f4b278a53d6b/scripts/environment/prepare.sh#L8-L10

JeanMertz commented:
> Unsure if changes to the other crates mentioned would ever cause the vrl playground to fail.
> 
> @JeanMertz @fuchsnj thoughts on this?

I'd love for it to build on any change in `lib/vrl`, as then you can quickly open the preview in a PR and see if whatever change you made to VRL works the way you expect it to work. E.g. say I update a diagnostic message, of course we have tests for all of this, but it's still nice occasionally to open the playground and see it working on there.

jszwedko commented on lib/vrl/web-playground/netlify.toml:
```diff
@@ -2,18 +2,16 @@
+  command = "cargo install --force --version 0.10.3 wasm-pack && wasm-pack build --target web --out-dir public"
 
+  # Ignore everything except the base directory and changes to lib/vrl/stdlib
+  ignore = "git diff --quiet $CACHED_COMMIT_REF $COMMIT_REF . ../stdlib"
```
I'd at least expand this to cover all of `lib/vrl` though `stdlib` has references to even more crates like `vector-common` so changes to these crates should also trigger a rebuild.

Personally I don't really think we need to skip builds.

JeanMertz commented:
> Personally I don't really think we need to skip builds.

That's a good point by @jszwedko. What triggered the need for this PR? I can see why we'd want to skip the build on any code that isn't involved in the VRL Playground, but if you track down dependencies, then I suspect you end up only ignoring `src/`, which contains Vector only code, and maybe a few `lib/` directories that aren't (yet) dependencies of the playground.

spencergilbert commented:
I suggested this as we're already exciting our build minutes on Netlify, as well as triggering failures on unrelated PRs - impacting other contributors. There isn't any reason to spend time/money rebuilding either site if the contents of the site haven't actually changed.

jonathanpv commented:
> I suggested this as we're already exciting our build minutes on Netlify, as well as triggering failures on unrelated PRs - impacting other contributors. There isn't any reason to spend time/money rebuilding either site if the contents of the site haven't actually changed.

Thinking about listening to changes to all of `lib` for now since as Jesse mentioned, changes to vector-common might impact vrl which impacts the playground.

I'll also add the bash script to check if wasm-pack is already installed. 

Although these are two separate tasks, they're small enough to keep in this PR. 
 
What do we think about this? @jszwedko @JeanMertz @spencergilbert

jszwedko commented:
> I suggested this as we're already exciting our build minutes on Netlify, as well as triggering failures on unrelated PRs - impacting other contributors. There isn't any reason to spend time/money rebuilding either site if the contents of the site haven't actually changed.

I think the tricky bit is figuring out _when_ things actually changed though. The VRL crates have dependencies on other `lib` crates. As Jean suggested, we could exclude `src` which should cut down a fair bit of deploys. We could restrict it to specific `lib` crates and just try to remember to update it when that changes.

@jonathanpv that sounds good to me as a next step.

jonathanpv commented:
I've added a few changes namely, I know it's bad practice to add multiple changes but these are so small I think it should be fine to include them in this PR
1. Improved and updated the readme
2. Included a script guard that checks if the correct wasm-pack version is installed
3. Updated and improved the index.html, debug messages, and also made any shared code automatically run, and also prevents the "run_vrl" function is not defined bug @fuchsnj might recall both these issues as it was mentioned in the slack

jonathanpv commented:
forgot to add the listen to lib directory as suggested in the above comments, just committed that one right now

jszwedko reviewed (approved):
Looks good! I like the addition of the examples to the readme.
